compiler/src/iree/compiler/Dialect/Flow/Transforms/ConvertConv2DToWinograd.cpp

```diff
@@ -0,0 +1,281 @@
+                                                    reassociations);
+  return rewriter.create<tensor::ExpandShapeOp>(loc, resultType, tensor,
+                                                reassociations);
+}
+
+namespace {
+
+class ConvertConv2DNhwcHwcf final
+    : public OpRewritePattern<linalg::Conv2DNhwcHwcfOp> {
+ public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::Conv2DNhwcHwcfOp convOp,
+                                PatternRewriter &rewriter) const override {
+    // Check that strides = 1
+    if (!hasAllOneValues(convOp.getStrides())) return failure();
+
+    // Check that dilations = 1
+    if (!hasAllOneValues(convOp.getDilations())) return failure();
+
+    // Check that kernel size = 3x3
+    Value kernel = convOp.getInputs()[1];
+    auto kernelType = kernel.getType().cast<ShapedType>();
+    if (!kernelType) return failure();
+    ArrayRef<int64_t> kernelShape = kernelType.getShape();
+    Type elementType = kernelType.getElementType();
+    const int64_t kh = kernelShape[0];
+    const int64_t kw = kernelShape[1];
+    if ((kh != 3) || (kw != 3)) return failure();
+    const int64_t kernelSize = kh;
+
+    // TODO: Make this a user-settable parameter once we have support
+    // for more tile sizes
+    const int64_t outputTileSize = 6;
+    const int64_t inputTileSize = outputTileSize + kernelSize - 1;
+
+    // Determine if filter transform can be constant folded (only true for
+    // inference)
+    Location loc = convOp.getLoc();
+    bool constantFoldFilter{false};
+    Value foldedKernel;
+    auto constOp = kernel.getDefiningOp<arith::ConstantOp>();
+    if (constOp) {
+      auto kernel = constOp.getValue().cast<DenseIntOrFPElementsAttr>();
+      if (!kernel) return failure();
+      ShapedType type = constOp.getType().cast<ShapedType>();
+      Type elementType = type.getElementType();
+      assert(elementType.isa<FloatType>());
+      ArrayRef<int64_t> shape = type.getShape();
+      DenseElementsAttr::iterator_range<APFloat> nonSplatValues =
+          kernel.getValues<APFloat>();
+      bool isSplat = kernel.isSplat();
+      float splatValue{0.0};
+      if (isSplat) {
+        splatValue = kernel.getSplatValue<APFloat>().convertToFloat();
+      }
+      SmallVector<int64_t> resultShape{inputTileSize * inputTileSize, shape[2],
+                                       shape[3]};
+      auto resultType = RankedTensorType::get(resultShape, elementType);
+      auto foldedKernelAttr =
+          foldFilterTransform(shape, inputTileSize, kernelSize, resultType,
+                              IREE::LinalgExt::Winograd::G_6x6_3x3, isSplat,
+                              splatValue, nonSplatValues, elementType);
+      foldedKernel = rewriter.replaceOpWithNewOp<arith::ConstantOp>(
+          constOp, foldedKernelAttr);
+      constantFoldFilter = true;
+    }
+    if (!constantFoldFilter) return failure();
+
+    // Create winograd input transform op
```

Comment:
Can we provide more details in the comment regarding the transformation done here? Like how do we transform the input, from what shape to what shape, etc. Links to details are fine. It would be a good reference for somebody else later to understand/modify the code easily.

Reply:
sure will add a detailed description